server/athenian/api/internal/features/entries.py

```diff
@@ -130,30 +130,77 @@ def _postprocess_cached_facts(
     return result
 
 
-class _PRMetricsLineGHCache:
-    @staticmethod
-    def postprocess(result: np.ndarray, *, metrics: Sequence[str], **kwargs: Any) -> np.ndarray:
-        # cache value has been stored with sorted(metrics), adapt before returning to caller
-        if (sorted_metrics := sorted(metrics)) != metrics:
-            sort_indexes = np.searchsorted(sorted_metrics, metrics)
-            return _sorted_result_metrics(result, sort_indexes)
+T = TypeVar("T")
 
-        return result
 
-    @staticmethod
-    def preprocess(result: np.ndarray, *, metrics: Sequence[str], **kwargs: Any) -> np.ndarray:
+class _CalcMetricsCache(Generic[T]):
+    """Cache processing for calculator methods having a list of metrics as argument.
+
+    Methods result must include a numpy array with elements ordered according to the order of
+    metrics in arguments.
+    Pre/post-processing functions in this class allows to share result cache across calls
+    with different metrics order.
+
+    """
+
+    @classmethod
+    def preprocess(cls, result: T, *, metrics: Sequence[str], **kwargs: Any) -> T:
         # cache value must be stored with sorted(metrics), adapt if needed
         if sorted(metrics) != metrics:
             sort_indexes = np.argsort(metrics)
-            return _sorted_result_metrics(result, sort_indexes)
+            return cls.sort_metric_values_in_result(result, sort_indexes)
+        return result
+
+    @classmethod
+    def postprocess(cls, result: T, *, metrics: Sequence[str], **kwargs: Any) -> T:
+        # cache value has been stored with sorted(metrics), adapt before returning it to the caller
+        if (sorted_metrics := sorted(metrics)) != metrics:
+            sort_indexes = np.searchsorted(sorted_metrics, metrics)
+            return cls.sort_metric_values_in_result(result, sort_indexes)
         return result
 
+    @classmethod
+    def sort_metric_values_in_result(cls, result: T, sort_indexes: Iterable[int]) -> T:
+        raise NotImplementedError()
 
-def _sorted_result_metrics(result: np.ndarray, metrics_sort_indexes: Iterable[int]) -> np.ndarray:
-    flat_copy = result.ravel().copy()
-    # each element of `copy` is a list of list of metrics
-    flat_copy[:] = [[[m[idx] for idx in metrics_sort_indexes] for m in obj] for obj in flat_copy]
-    return flat_copy.reshape(result.shape)
+    @classmethod
+    def _sorted_metric_values(cls, values: np.ndarray, sort_indexes: Iterable[int]) -> np.ndarray:
+        flat_copy = values.ravel().copy()
+        # each element of `flat_copy` is a list of list of metrics
+        flat_copy[:] = [[[m[idx] for idx in sort_indexes] for m in obj] for obj in flat_copy]
```

Comment:
```suggestion
        flat_copy[:] = ([[m[idx] for idx in sort_indexes] for m in obj] for obj in flat_copy)
```
performance: avoid materializing the list twice

Reply:
It seems it is  not  supported to assign from an iterator,